framework/src/base/Adaptivity.C

```diff
@@ -221,21 +222,88 @@ Adaptivity::adaptMesh(std::string marker_name /*=std::string()*/)
     _mesh_refinement->make_flags_parallel_consistent();
 
   // Perform refinement and coarsening
+  auto & lm_mesh = _mesh_refinement->get_mesh();
+  const auto old_removal_flag = lm_mesh.allow_remote_element_removal();
+  const auto old_renumbering_flag = lm_mesh.allow_renumbering();
+  if (_displaced_problem)
+  {
+    // If we have both reference and displaced meshes, then we have to be very careful to keep them
+    // in sync. We have proxy relationship managers that rely on matching ghosted elements via
+    // unique id, so given that, we must make sure that we do not try to delete elements or renumber
+    // them (changing their unique id) at the stage that only one mesh has been refined. We must
```

Comment:
Wait, what had been going on here?  Mesh renumbering should never change unique_id() on an already-existing object, only id().

Reply:
I got scared when I saw this code in `DistributedMesh::renumber_dof_objects` and saw `set_unique_id` but I admittedly did not look closely at the surrounding code
```c++
#ifdef LIBMESH_ENABLE_UNIQUE_ID
  auto unique_gather_functor =
    [
#ifndef NDEBUG
     this,
#endif
     &objects]
    (processor_id_type, const std::vector<dof_id_type> & ids,
     std::vector<unique_id_type> & data)
    {
      std::size_t ids_size = ids.size();
      data.resize(ids_size);

      for (std::size_t i=0; i != ids_size; ++i)
        {
          T * obj = objects[ids[i]];
          libmesh_assert(obj);
          libmesh_assert_equal_to (obj->processor_id(), this->processor_id());
          data[i] = obj->valid_unique_id() ? obj->unique_id() : DofObject::invalid_unique_id;
        }
    };

  auto unique_action_functor =
    [&objects]
    (processor_id_type libmesh_dbg_var(pid),
     const std::vector<dof_id_type> & ids,
     const std::vector<unique_id_type> & data)
    {
      for (auto i : index_range(ids))
        {
          T * obj = objects[ids[i]];
          libmesh_assert (obj);
          libmesh_assert_equal_to (obj->processor_id(), pid);
          if (!obj->valid_unique_id() && data[i] != DofObject::invalid_unique_id)
            obj->set_unique_id(data[i]);
        }
    };

  const unique_id_type * unique_ex = nullptr;
  Parallel::pull_parallel_vector_data
    (this->comm(), requested_ids, unique_gather_functor,
     unique_action_functor, unique_ex);
#endif
```
So historically we have not added these `ProxyRelationshipManagers` to the mesh (e.g. as geometric ghosting functors), but now I want to. After adding them to the mesh, I needed to at a minimum prevent remote element deletion when the reference mesh has been refined and the displaced mesh has not. Perhaps disabling renumbering is not required if we're not modifying existing unique ids 🤔

Reply:
Ah you don't have to look very closely at all do you? Lol
```c++
         if (!obj->valid_unique_id() && data[i] != DofObject::invalid_unique_id)
            obj->set_unique_id(data[i]);
```

Reply:
That's some old code there ... I think it might be redundant now, actually - we ought to currently have every ghost's unique_id in sync earlier in the process of modifying a mesh.  Once you've got some good test coverage for the kinds of tricky messes fixed by this PR, let me know.  I may want to try out a libMesh PR replacing that "update any object that doesn't already have a valid unique_id" with "scream and die if any object doesn't already have a valid unique_id", and just see what happens.

Reply:
prod-to-resolve-button @roystgnr :smile:

Reply:
Ha!  Sorry, yeah, this PR is good to go.  I just need to remember to make that PR in libMesh, see if I'm right about the code there having been made redundant or if there's some special case I'm forgetting.